### ISSUE TITLE
enviroment variable set in script not available in dependent tasks

### ISSUE THREAD

**Issue by x37v** [sagiegurari/cargo-make#275]
**Describe the bug**
I have a relatively complex scenario for setting some env vars that I want to use in my build tasks. I'm using a rust script and `env::set_var` but the variables I've set are not visible to my dependent tasks.

**To Reproduce**
1. make a `Makefile.toml` file with the contents from below.
2. run `cargo make echo`

**Error Stack**
```console
alex@shred:~/projects/rust/puredata/macros/examples/helloworld$ cargo make echo
[cargo-make] INFO - cargo make 0.22.0
[cargo-make] INFO - Using Build File: Makefile.toml
[cargo-make] INFO - Task: echo
[cargo-make] INFO - Profile: development
[cargo-make] INFO - Running Task: init
[cargo-make] INFO - Execute Command: "cargo" "script" "/tmp/cargo-make/qVL5LcjQCP.rs"
[cargo-make] INFO - Running Task: echo
PLATFORM_DYLIB_EXT:
PLATFORM_EXTERNAL_EXT:
[cargo-make] INFO - Running Task: end
[cargo-make] INFO - Build Done in 0 seconds.
```

**Code Sample**
```rust
[tasks.init]
script_runner = "@rust"
script = [
'''
use std::env;
fn main() -> std::result::Result<(), std::env::VarError> {
    let os = env::var("CARGO_MAKE_RUST_TARGET_OS")?;
    let arch = env::var("CARGO_MAKE_RUST_TARGET_ARCH")?;

    let (prefix, dylib_suffix) = match os.as_ref() {
        "linux" | "freebsd" => ("l", "so"),
        "macos" => ("d", "dylib"),
        "windows" => ("m", "dll"),
        "openbsd" => ("o", "so"),
        _ => panic!("unhandled os {}", os)
    };

    let suffix = match arch.as_ref() {
        "x86" => "i386",
        "x86_64" => "amd64",
        "arm" => "arm",
        "aarch64" => "arm64",
        "powerpc" => "ppc",
        _ => panic!("unhandled arch {}", arch)
    };

    env::set_var("PLATFORM_EXTERNAL_EXT", format!("{}_{}", prefix, suffix));
    env::set_var("PLATFORM_DYLIB_EXT", dylib_suffix);

    Ok(())
}
'''
]

[tasks.echo]
script = [
'''
echo PLATFORM_DYLIB_EXT: ${PLATFORM_DYLIB_EXT}
echo PLATFORM_EXTERNAL_EXT: ${PLATFORM_EXTERNAL_EXT}
'''
]

```

**Comment by sagiegurari** [comment]
commands and scripts are invoked as child processes and therefore do not change the env of the main cargo-make parent process.
there are [other ways to change the env](https://github.com/sagiegurari/cargo-make#usage-env)

Having said that, looking at your use case, it does seem a nice feature to have a way to map env values to other env values. meaning to do something like the decoding you are doing in your example, just by expressing it via toml.

i'll think of something and update here.

**Comment by x37v** [comment]
@sagiegurari yeah.. i figured that was the case and wasn't sure if I should log this as a bug or a feature request.

The [other wasy to change the env](https://github.com/sagiegurari/cargo-make#usage-env) are a bit too cumbersome for this more complex scenario.. I guess I could use a script but the multiline on 1 line with `\n` didn't seem great.. if at least i could evaluate this rust code and print a single line from there for each of these values, that would work.

```
[env]
PLATFORM_EXTERNAL_EXT = {
script_runner = "@rust",
script = [
'''
use std::env;
fn main() -> std::result::Result<(), std::env::VarError> {
    let os = env::var("CARGO_MAKE_RUST_TARGET_OS")?;
    let arch = env::var("CARGO_MAKE_RUST_TARGET_ARCH")?;

    let (prefix, dylib_suffix) = match os.as_ref() {
        "linux" | "freebsd" => ("l", "so"),
        "macos" => ("d", "dylib"),
        "windows" => ("m", "dll"),
        "openbsd" => ("o", "so"),
        _ => panic!("unhandled os {}", os)
    };

    let suffix = match arch.as_ref() {
        "x86" => "i386",
        "x86_64" => "amd64",
        "arm" => "arm",
        "aarch64" => "arm64",
        "powerpc" => "ppc",
        _ => panic!("unhandled arch {}", arch)
    };

    println!("{}_{}", prefix, suffix);
    Ok(())
}
'''
]
}
```

OR, if there were overrides for `TARGET_ARCH` and `TARGET_OS`

something like:

```
[env.target_os.linux]
PLATFORM_EXTERNAL_EXT_PREFIX = "l"

[env.target_arch.x86_64]
PLATFORM_EXTERNAL_EXT_SUFFIX = "amd64"

[task.somethinglater]
PLATFORM_EXTERNAL_EXT = "${PLATFORM_EXTERNAL_EXT_PREFIX}_${PLATFORM_EXTERNAL_EXT_SUFFIX}"
```

**Comment by sagiegurari** [comment]
I was thinking of enabling 'decode' for env vars like the following:

```toml
[env]
PLATFORM_EXTERNAL_EXT_SUFFIX={decode={source = "${CARGO_MAKE_RUST_TARGET_OS}", default = "", mapping = {"linux" = "so", "macos" = "dylib", "windows" = "dll", "openbsd" = "so"}}}
```

so source would be the value to decode and could accept env vars for example.
default is the output of PLATFORM_EXTERNAL_EXT_SUFFIX in case no mapping was found.
mapping holds source value to target value.

**Comment by x37v** [comment]
@sagiegurari `decode` sounds like it would work well..
might want to allow an error if there is no match.. though I guess you have other operations that can achieve that assertion.

**Comment by sagiegurari** [comment]
@x37v you can now try out the new decode capability in the 0.22.1 development branch.
you can look at the [docs](https://github.com/sagiegurari/cargo-make/tree/0.22.1#usage-env-config)

```toml
[env]
LIBRARY_EXTENSION = { source = "${CARGO_MAKE_RUST_TARGET_OS}", default_value = "unknown", mapping = {"linux" = "so", "macos" = "dylib", "windows" = "dll", "openbsd" = "so" } }
```

you don't need to write rust code anymore, just define it in the toml.
as for errors, i do not believe its up to env setup to do that, but you can do validations in your task using conditions.

**Comment by x37v** [comment]
@sagiegurari works for me on linux with my minimal testing. looks great!

**Comment by sagiegurari** [comment]
thanks for checking

**Comment by sagiegurari** [comment]
@x37v this is now available in 0.22.1 version which was just published.
i think you can do all your mapping now without rust code.
